### ISSUE TITLE
PushbackInputStream degrading performance.

### ISSUE THREAD

**Issue by ravenchild** [NanoHttpd/nanohttpd#228]
I've recently moved some old code to use the latest Nanohttpd. I was getting ~2.5k requests/sec before the switch and now I'm getting ~100 requests/sec. 

I tried to profile with visualvm but couldn't narrow it down to a specific cause. I then used method level tracing and found the PushbackInputStream was spending a good amount of time (90% of execute()) in PushbackInputStream.read().  I've built my code in such a way that the Nanohttpd file is unmodified. My serve() function is using 0.8% of execute()'s time.

Is there any effort going into optimizing the reads?

**Comment by ritchieGitHub** [comment]
from now on yes! thanks for the analysis. could you specify your configuration? hw/os/java-version
could you provide your test implementation ?

**Comment by ravenchild** [comment]
Tried with on RPi, Mac, Linux, Android so the problem seems to affect each. I looked a little closer at the issue last night and realized the socket read is what is taking all of the time. My previous code was using a slightly optimized nanohttpd 1.25 and the way reading the headers are read has changed completely.

**Comment by ravenchild** [comment]
I wrapped the socket input stream in a BufferedInputStream and got a ~60% improvement in speed. Still not as good as it was.

**Before BufferedInputStream:**
```
Server Software:        
Server Hostname:        localhost
Server Port:            8081
SSL/TLS Protocol:       TLSv1,EDH-RSA-DES-CBC3-SHA,2048,168

Document Path:          /rest/infostring
Document Length:        86 bytes

Concurrency Level:      20
Time taken for tests:   32.359 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      222000 bytes
HTML transferred:       86000 bytes
Requests per second:    30.90 [#/sec] (mean)
Time per request:       647.187 [ms] (mean)
Time per request:       32.359 [ms] (mean, across all concurrent requests)
Transfer rate:          6.70 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       27  451 595.2    237    3655
Processing:     1  195 169.4    163    2978
Waiting:        0   53 125.2     28    2473
Total:         85  647 633.9    403    5586

Percentage of the requests served within a certain time (ms)
  50%    403
  66%    502
  75%    596
  80%    694
  90%   1479
  95%   2077
  98%   3038
  99%   3180
 100%   5586 (longest request)
```

**After BufferedInputStream:**
```
Server Software:        
Server Hostname:        localhost
Server Port:            8081
SSL/TLS Protocol:       TLSv1,EDH-RSA-DES-CBC3-SHA,2048,168

Document Path:          /rest/infostring
Document Length:        86 bytes

Concurrency Level:      20
Time taken for tests:   20.104 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      221000 bytes
HTML transferred:       86000 bytes
Requests per second:    49.74 [#/sec] (mean)
Time per request:       402.074 [ms] (mean)
Time per request:       20.104 [ms] (mean, across all concurrent requests)
Transfer rate:          10.74 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       27  252 114.2    258     869
Processing:    16  148  98.1    119     604
Waiting:        0   41  48.4     25     549
Total:        167  399 106.4    380     951

Percentage of the requests served within a certain time (ms)
  50%    380
  66%    401
  75%    423
  80%    439
  90%    515
  95%    584
  98%    880
  99%    887
 100%    951 (longest request)
```

**Comment by elonen** [comment]
Are you serving GET or POST (or PUT) requests? HTTPS?

**Comment by ravenchild** [comment]
GET requests over HTTPS. I return a simple JSON string and I don't have any special headers on the request. Apache Benchmark was started with ```ab -c 20 -n 1000```.

**Comment by ritchieGitHub** [comment]
This weekend, I did not have the time. but in the next block I will follow this issue

**Comment by ritchieGitHub** [comment]
I think I have a suspect ... can you confirm that the CPU is not the problem but blocking reads? the read should not try to fill the whole buffer but use the max(1,min (bufferssize,available() )) Next weekend I can try it but If you @ravenchild can it would help save a lot of time.

**Comment by ravenchild** [comment]
I've got some time tonight. Will try to look into it some more. I added the extra buffer and forgot about this issue haha.

**Comment by ravenchild** [comment]
I re-ran the benchmarks and got ~60 requests per second for SSL and 1000+ requests per second on the non-SSL. I'm guessing the laptop I'm working on (2008 era core2duo) can't compete with my server but the changes in my pull request help (~70% improvements). 

I don't know if I broke anything or if reseting and skipping the correct way to handle reading the headers. For all its worth, it works for my service implementation.

I tried rewriting the decode headers to prevent the initial read through. I wasn't able to do this and keep rlen/splitbyte accurate for all cases. It would be great if we didn't have to read the entire header beforehand but buffering seems to be the correct solution.
